otaclient/app/downloader.py

```diff
@@ -158,10 +161,25 @@ class ZstdDecompressionAdapter(DecompressionAdapterProtocol):
     def __init__(self) -> None:
         self._dctx = zstandard.ZstdDecompressor()
 
-    def iter_chunk(self, src_stream: Union[IO[bytes], ByteString]) -> Iterator[bytes]:
+    def iter_chunk(self, src_stream) -> Iterator[bytes]:
         yield from self._dctx.read_to_iter(src_stream)
 
 
+class _HTTPResponseWrapper:
+    def __init__(self, resp: urllib3.HTTPResponse) -> None:
+        self._resp = resp
+        self._bytes_read = 0
+        self.retry_count = len(resp.retries.history) if resp.retries else 0
+
+    def read(self, size: int) -> bytes:
+        data = self._resp.read(size)
```

Comment:
@Bodong-Yang Let me confirm, if the file size is 2GB, 2GB is loaded onto the main memory?

Reply:
Ah, sorry, this just read chunk size, not whole size.

Reply:
This wrapper is only used by decompressor's `iter_chunk` method, and decompressor will only read a chunk of data each time from this wrapper. (The read size is `zstandard.DECOMPRESSION_RECOMMENDED_INPUT_SIZE`, the default recommended value of zstandard's `read_to_iter` stream decompressing method).

Reply:
Basically this wrapper is a stub for raw response, wraps the raw response's read method, tracking the actual number of bytes we read from the raw response.